torch/optim/decoupled_weight_decay_optim.py

```diff
@@ -0,0 +1,99 @@
+import torch
+import warnings
+from .adam import Adam
+from .sgd import SGD
+
+
+def extend_with_decoupled_weight_decay(base_optimizer):
+    r"""Extends ```base_optimizer``` with decoupled weight decay.
+
+    This factory function returns a new optimizer class, in which the weight
+    decay is decoupled from the optimization steps w.r.t. to the loss function,
+    as proposed in Decoupled Weight Decay Regularization by Loshchilov & Hutter:
+    https://arxiv.org/abs/1711.05101
+
+    For SGD variants, this simplifies hyperparameter search since it decouples
+    the settings of weight decay and learning rate.
+    For adaptive gradient algorithms, it regularizes variables with large
+    gradients more than L2 regularization would, which was shown to yield better
+    training loss and generalization error in the paper above.
+    
+    .. note:: The weight decay is not coupled with the learning rate. If a
+        scheduler is used for the learning rate, it is reasonable to schedule
+        the weight decay as well.
+
+    .. note:: To prevent copying of the parameters, the weight decay is
+        applied before the optimizer update. If the update step relies on
+        the value of the parameter (and not just the gradient), this function
+        should not be used.
+
+    Example:
+        >>> AdamW = extend_with_decoupled_weight_decay(torch.optim.Adam)
+        >>> optimizer = AdamW(decoupled_weight_decay=1e-5, lr=1e-3)
+    """
+    
+    class DecoupledWeightDecayOptimizer(base_optimizer):
+
+        __doc__ = \
+        """
+        Implements the {0}W algorithm.
+            
+        I.e., this class implements the update step of {0} with weight
+        decay, where the weight decay is decoupled from the optimization
+        steps w.r.t. to the loss function, as proposed in Decoupled Weight
+        Decay Regularization by  Loshchilov & Hutter:
+        https://arxiv.org/abs/1711.05101
+
+        .. note:: The weight decay is not coupled with the  learning rate.
+            If a scheduler is used for the learning  rate, it is reasonable to
+            schedule the weight decay as well.
+        
+        Arguments:
+            decoupled_weight_decay (float, optional): weight decay (default: 0)
+            *args: arguments passed to the base optimizer.
+            **kwargs: keyword arguments passed to the base optimizer.
+        
+        Details on the update step can be found in the documentation of the
+        wrapped {0} optimizer.
+        """.format(base_optimizer.__name__)
+
+        def __init__(self, decoupled_weight_decay, *args, **kwargs):
+            super(DecoupledWeightDecayOptimizer, self).__init__(*args, **kwargs)
+            if self.defaults["weight_decay"] != 0:
+                warnings.warn(
```

Comment:
Do people want to do both at the same time? I suspect people understand AdamW and SGDW as not having L2 anymore. If so, I'd suggest to simply biggy back on the parameter provided in Adam. There shouldn't be confusion when someone invokes AdamW which regularization they want.

Reply:
I've changed the the input argument to weight_decay. Internally however, it still needs to be called differently, otherwise the optimizers' update steps will do additional L2 decay.

Reply:
On a second thought, if general schedules are on their way, it might be confusing having to schedule "decoupled_weight_decay", while the input argument is called "weight decay"?